[XLA] Speed up col reduction for dtype smaller then 32 bits.

nouiz opened the pull request:
It was optimized for float32. Now that we frequently use f16, bf16 and sometimes int8, we need to update it.
This speed up the following kernel on V100 from 860us to 631us.

```
region_0.18.clone {
  Arg_0.0 = f32[] parameter(0)
  Arg_1.0 = f32[] parameter(1)
  ROOT add.0 = f32[] add(Arg_0.0, Arg_1.0)
}
fused_computation.1 {
  param_0.4 = bf16[32,16,512,512]{3,2,1,0} parameter(0)
  convert.31 = f32[32,16,512,512]{3,2,1,0} convert(param_0.4)
  constant_1 = f32[] constant(0)
  ROOT reduce.0 = f32[16,512,512]{2,1,0} reduce(convert.31, constant_1), dimensions={0}, to_apply=region_0.18.clone
}
ENTRY main {
 Arg_3.4 = bf16[32,16,512,512]{3,2,1,0} parameter(0)
 ROOT fusion.1 = f32[16,512,512]{2,1,0} fusion(Arg_3.4), kind=kInput, calls=fused_computation.1
}
```

@cheshire

nouiz committed:
[XLA] Speed up col reduction for dtype smaller then 32 bits.

nouiz commented:
Any update? It was approved 3 days ago, but isn't merged yet.

cheshire commented:
@gflegar

cheshire commented:
//tensorflow/compiler/xla/service/gpu/tests:gpu_kernel_tiling_test is failing, is it passing for you?

nouiz commented:
I fixed the test failure by removing the test change in my commit.
Should be good now.

cheshire commented:
That would be untested code though. Sorry for us lagging behind with Ampere CI. Could you predicate on compute capability at runtime? We do have a number of tests doing that.

nouiz commented:
> That would be untested code though. Sorry for us lagging behind with Ampere CI. Could you predicate on compute capability at runtime? We do have a number of tests doing that.

I rebased and added a test. I already modified one existing tests, but an older rebase removed it.
